File: infra/supabase/functions/_shared/planner/steps/state_machine_config.ts
import type { Dimension, SkillStatus } from '../models/enums.ts';

/**
 * Start value and cap for the spaced-repetition interval of one skill
 * status. In days.
 */
export interface IntervalConfig {
  readonly start: number;
  readonly cap: number;
}

/**
 * The slice of `content/planer.yaml` the state machine needs (sections
 * `spaced_repetition` and `stufen`). Passed in, not imported (CLAUDE.md,
 * rule 1/10) — loading the YAML file belongs elsewhere, not in
 * `_shared/planner/`.
 */
export interface StateMachineConfig {
  /** 3× „succeeded" at the current level raises one dimension. */
  readonly increaseAfterSuccesses: number;

  /** 2× „not yet" in a row lowers the active dimension. */
  readonly decreaseAfterFailures: number;

  /** Order in which dimensions are raised: duration → distance → distraction. */
  readonly order: readonly Dimension[];

  readonly generalizeAtDistraction: number;

  /** Multiplier applied to the interval on „succeeded". */
  readonly successFactor: number;

  /**
   * Only for the statuses that have their own row in `content/planer.yaml`:
   * building, generalizing, consolidated, maintenance.
   */
  readonly intervals: ReadonlyMap<SkillStatus, IntervalConfig>;
}

File: packages/engine/lib/src/planner/state_machine_config.dart
import 'package:meta/meta.dart';

import '../models/enums.dart';

/// Start value and cap for the spaced-repetition interval of one skill
/// status. In days.
@immutable
class IntervalConfig {
  const IntervalConfig({required this.start, required this.cap});

  final int start;
  final int cap;
}

/// The slice of `content/planer.yaml` the state machine needs (sections
/// `spaced_repetition` and `stufen`). Passed in, not imported (CLAUDE.md,
/// rule 10) — loading the YAML file belongs in the `tool` package, not the
/// engine.
@immutable
class StateMachineConfig {
  const StateMachineConfig({
    required this.increaseAfterSuccesses,
    required this.decreaseAfterFailures,
    required this.order,
    required this.generalizeAtDistraction,
    required this.successFactor,
    required this.intervals,
  });

  /// 3× „succeeded" at the current level raises one dimension.
  final int increaseAfterSuccesses;

  /// 2× „not yet" in a row lowers the active dimension.
  final int decreaseAfterFailures;

  /// Order in which dimensions are raised: duration → distance →
  /// distraction.
  final List<Dimension> order;

  final int generalizeAtDistraction;

  /// Multiplier applied to the interval on „succeeded".
  final double successFactor;

  /// Only for the statuses that have their own row in `content/planer.yaml`:
  /// building, generalizing, consolidated, maintenance.
  final Map<SkillStatus, IntervalConfig> intervals;
}
